wire-library/wire-gradle-plugin/src/main/kotlin/com/squareup/wire/gradle/WirePlugin.kt

```diff
@@ -45,104 +51,178 @@ class WirePlugin : Plugin<Project> {
       outputs.add(JavaOutput())
     }
 
+    val hasJavaOutput = outputs.any { it is JavaOutput }
+    val hasKotlinOutput = outputs.any { it is KotlinOutput }
+    check(!hasKotlinOutput || kotlin.get()) {
+      "Wire Gradle plugin applied in " +
+          "project '${project.path}' but no supported Kotlin plugin was found"
+    }
+
+    addWireRuntimeDependency(hasJavaOutput, hasKotlinOutput)
+
     for (output in outputs) {
       if (output.out == null) {
         output.out = "${project.buildDir}/generated/source/wire"
       } else {
         output.out = project.file(output.out!!).path
       }
     }
+    val generatedSourcesDirectories = outputs.map { output -> File(output.out!!) }.toSet()
+    // TODO(benoit) Do we want to delete all output directories? It creates problem if protos are
+    //  generated in a shared folder BUT there would no be any forgotten code left over code when
+    //  the proto schema shrinks.
+    // for (generatedSourcesDirectory in generatedSourcesDirectories) {
+    //   generatedSourcesDirectory.deleteRecursively()
+    // }
+
+    val common = sources.singleOrNull { it.type == KotlinPlatformType.common }
+    for (generatedSourcesDirectory in generatedSourcesDirectories) {
+      common?.sourceDirectorySet
+          ?.srcDir(generatedSourcesDirectory.toRelativeString(project.projectDir))
+    }
+    sources.forEach { source ->
+      if (common == null) {
+        for (generatedSourcesDirectory in generatedSourcesDirectories) {
+          source.sourceDirectorySet
+              .srcDir(generatedSourcesDirectory.toRelativeString(project.projectDir))
+        }
+      }
 
-    val targets = outputs.map { it.toTarget() }
-    val wireTask = project.tasks.named("generateProtos") as TaskProvider<WireTask>
-    wireTask.configure {
-      if (it.logger.isDebugEnabled) {
-        sourceInput.debug(it.logger)
-        protoInput.debug(it.logger)
+      val sourceInput = WireInput(project.configurations.named("protoSource"))
+      if (extension.sourcePaths.isNotEmpty() ||
+          extension.sourceTrees.isNotEmpty() ||
+          extension.sourceJars.isNotEmpty()) {
+        sourceInput.addTrees(project, extension.sourceTrees)
+        sourceInput.addJars(project, extension.sourceJars)
+        sourceInput.addPaths(project, extension.sourcePaths)
+      } else {
+        sourceInput.addPaths(project, setOf("src/main/proto"))
+      }
+
+      val protoInput = WireInput(project.configurations.named("protoPath"))
+      if (extension.protoPaths.isNotEmpty() ||
+          extension.protoTrees.isNotEmpty() ||
+          extension.protoJars.isNotEmpty()) {
+        protoInput.addTrees(project, extension.protoTrees)
+        protoInput.addJars(project, extension.protoJars)
+        protoInput.addPaths(project, extension.protoPaths)
+      }
+
+      val targets = outputs.map { it.toTarget() }
+      val task =
+          project.tasks.register("generate${source.name.capitalize()}Protos",
```

Comment:
i don't think we should using sourceSets in task names.

a Gradle build target maps to a Gradle variant (which I believe inspired by the Android variant system).  

The variant helps decide what source sets to use, as a sort of implementation detail.

Reply:
think assemble -> assembleDebug, assembleRelease, not assembleMain.

Reply:
There would not be assembleMain on Android, only Debug and Release.

Reply:
Would a task in a plain Kotlin JVM project be called `generateMainProtos` or `generateProtos`? My preference is for the latter.

Reply:
`generateProtos` would always be there whatever setup there is. It will execute all sources tasks.
Sor a plain JVM project would have
`generateProtos` and `generateMainProtos`.